File: src/main/kotlin/nl/juraji/albums/domain/PictureTagsService.kt
package nl.juraji.albums.domain

import nl.juraji.albums.domain.pictures.PictureTagsRepository
import nl.juraji.albums.domain.tags.Tag
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class PictureTagsService(
    private val pictureTagsRepository: PictureTagsRepository
) {
    fun getPictureTags(pictureId: String): Flux<Tag> =
        pictureTagsRepository.findPictureTags(pictureId)

    fun addTagToPicture(pictureId: String, tagId: String): Mono<Tag> =
        pictureTagsRepository.addTagToPicture(pictureId, tagId)

    fun removeTagFromPicture(pictureId: String, tagId: String): Mono<Void> =
        pictureTagsRepository.removeTagFromPicture(pictureId, tagId)
}

File: src/web/app/@services/picture-tags.service.ts
import {Injectable} from '@angular/core';
import {environment} from '@environment';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class PictureTagsService {
  private readonly baseUri = `${environment.apiBaseUri}/pictures`;

  constructor(private readonly httpClient: HttpClient) {
  }

  getPictureTags(pictureId: string): Observable<Tag[]> {
    return this.httpClient.get<Tag[]>(`${this.baseUri}/${pictureId}/tags`);
  }

  addTagToPicture(pictureId: string, tag: Tag): Observable<Tag> {
    return this.httpClient.post<Tag>(`${this.baseUri}/${pictureId}/tags`, tag);
  }

  removeTagFromPicture(pictureId: string, tagId: string): Observable<void> {
    return this.httpClient.delete<void>(`${this.baseUri}/${pictureId}/tags/${tagId}`);
  }
}
